include/range/v3/utility/iterator.hpp

```diff
@@ -591,49 +570,55 @@ namespace ranges
-                std::basic_ostream<Char, Traits> *sout_;
-                Char const *delim_;
-                template<typename U,
-                    typename V = meta::if_<std::is_void<T>, U, T>,
-                    meta::if_<std::is_convertible<U, V const &>, int> = 0>
-                proxy &operator=(U &&t)
+                std::basic_ostream<Char, Traits> *sout_ = nullptr;
+                Char const *delim_ = nullptr;
+
+                struct mixin : protected basic_mixin<ostream_cursor>
+                {
+                    using char_type = Char;
+                    using traits_type = Traits;
+
+                    using basic_mixin<ostream_cursor>::basic_mixin;
+
+                    mixin() = default;
+                    explicit constexpr mixin(std::basic_ostream<Char, Traits> &sout) noexcept
+                    : basic_mixin<ostream_cursor>(ostream_cursor{sout})
+                    {}
+                    constexpr mixin(std::basic_ostream<Char, Traits> &sout,
+                        Char const *delim) noexcept
+                    : basic_mixin<ostream_cursor>(ostream_cursor{sout, delim})
+                    {}
+                };
+
+                ostream_cursor() = default;
+                explicit constexpr ostream_cursor(std::basic_ostream<Char, Traits> &sout) noexcept
+                  : sout_(std::addressof(sout))
+                {}
+                constexpr ostream_cursor(std::basic_ostream<Char, Traits> &sout,
+                    Char const *delim) noexcept
+                  : sout_(std::addressof(sout)), delim_(delim)
+                {}
+
+                template<typename U = T,
+                    CONCEPT_REQUIRES_(Same<T, U>() || Same<T, void>())>
+                void set(const U& u)
                 {
                     RANGES_ASSERT(sout_);
-                    *sout_ << static_cast<V const &>(t);
+                    *sout_ << u;
                     if(delim_)
                         *sout_ << delim_;
-                    return *this;
                 }
             };
-        public:
-            using difference_type = std::ptrdiff_t;
-            using char_type = Char;
-            using traits_type = Traits;
-            ostream_iterator() = default;
-            ostream_iterator(std::basic_ostream<Char, Traits> &sout, Char const *delim = nullptr) noexcept
-              : sout_(&sout), delim_(delim)
-            {}
-            proxy operator*() const noexcept
-            {
-                return {sout_, delim_};
-            }
-            ostream_iterator &operator++()
-            {
-                return *this;
-            }
-            ostream_iterator operator++(int)
-            {
-                return *this;
-            }
-        };
+        }
+        /// \endcond
+
+        template<typename T = void, typename Char = char, typename Traits = std::char_traits<Char>>
+        using ostream_iterator = basic_iterator<detail::ostream_cursor<T, Char, Traits>>;
```

Comment:
> Why? Just curious.

Because (1) I habitually constrain assignment operators with `&`, and (2) I broke the magic on `set`/`write` that made converting assignments work. This is now fixed on cmcstl2 and in this PR.